client/src/connection/rest/job.ts

```diff
@@ -0,0 +1,170 @@
+import { Compute, getApiConfig, stateOptions } from "./common";
+import {
+  LogsApi,
+  JobsApi,
+  JobsApiGetJobStateRequest,
+  Job,
+  Result,
+  ResultCollection,
+  LogLine,
+  Link,
+} from "./api/compute";
+import { AxiosRequestConfig, AxiosResponse } from "axios";
+
+export class ComputeJob extends Compute {
+  api;
+  logs;
+  _self: Job;
+
+  constructor(id: string) {
+    super();
+
+    //set the id
+    this._self = { id: id };
+
+    //Get the apis I need
+    this.api = JobsApi(getApiConfig());
+    this.logs = LogsApi(getApiConfig());
+  }
+
+  get id(): string {
+    return this._self?.id || "";
+  }
+
+  static fromInterface(job: Job): ComputeJob {
+    const obj = new ComputeJob("");
+    obj._self = job;
+    return obj;
+  }
+
+  static fromResponse(response: AxiosResponse): ComputeJob {
+    const obj = ComputeJob.fromInterface(response.data);
+    obj.etag = response.headers.etag;
+    return obj;
+  }
+
+  async followLink<T>(
+    linkName: string,
+    options?: AxiosRequestConfig
+  ): Promise<AxiosResponse<T>> {
+    if (this._self.links === undefined) await this.self();
+    const link = this.getLink(this._self.links, linkName);
+    if (link === undefined)
+      throw new Error(`Job does not have '${linkName}' link`);
+
+    return await this.requestLink(link, options);
+  }
+
+  async getState(options?: stateOptions): Promise<string> {
+    const parms: JobsApiGetJobStateRequest = {
+      sessionId: this._self.sessionId,
+      jobId: this.id,
+      wait: options?.wait,
+      ifNoneMatch: options?.onChange ? this.etag : undefined,
+    };
+
+    const resp = await this.api.getJobState(parms);
+    if (resp.status === 200) {
+      //Set the new etag
+      this.etag = resp.etag;
+
+      //return the state
+      return resp.data;
+    } else if (resp.status === 304) {
+      //Not modified
+      return await this.getState(); //This is bad. We need to cache the last state value
```

Comment:
Is more work needed, here?

Reply:
Are there scenarios where this could lean to endless recursion? (i.e. getState calls itself over and over)

Reply:
Just copy/pasting from a different comment for visibility...
> The getState function can never lead to recursion. The only way a 304 is returned is if the if-none-match header is sent. calling getState with no options, will send a simple call to getState which simply returns the current state of the job.